ironfish/src/wallet/wallet.ts

```diff
@@ -565,35 +566,77 @@ export class Wallet {
   private async upsertAssetsFromDecryptedNotes(
     account: Account,
     decryptedNotes: DecryptedNote[],
-    blockHeader?: WalletBlockHeader,
+    blockHeader: WalletBlockHeader,
     tx?: IDatabaseTransaction,
   ): Promise<void> {
     for (const { serializedNote } of decryptedNotes) {
       const note = new Note(serializedNote)
-      const asset = await this.walletDb.getAsset(account, note.assetId(), tx)
+      const asset = await this.getOrBackfillAsset(account, note.assetId(), tx)
+      Assert.isNotNull(asset, 'Asset must be non-null in the chain')
+      await account.updateAssetWithBlockHeader(asset, blockHeader, tx)
+    }
+  }
 
-      if (!asset) {
-        const chainAsset = await this.getChainAsset(note.assetId())
-        Assert.isNotNull(chainAsset, 'Asset must be non-null in the chain')
-        await account.saveAssetFromChain(
-          chainAsset.createdTransactionHash,
-          chainAsset.id,
-          chainAsset.metadata,
-          chainAsset.name,
-          chainAsset.nonce,
-          chainAsset.creator,
-          chainAsset.owner,
-          blockHeader,
-          tx,
-        )
-      } else if (blockHeader) {
-        await account.updateAssetWithBlockHeader(
-          asset,
-          { hash: blockHeader.hash, sequence: blockHeader.sequence },
-          tx,
-        )
+  /**
+   * Ensures that the wallet db contains information about the assets involved
+   * in the given notes and mints.
+   *
+   * This method checks that each asset is in the wallet db and, if it cannot
+   * be found, then the information is copied from the chain db into the wallet
+   * db.
+   */
+  private async backfillAssets(
+    account: Account,
+    decryptedNotes: DecryptedNote[],
+    mints: MintDescription[],
+    tx?: IDatabaseTransaction,
+  ): Promise<void> {
+    const assetIds: Set<Buffer> = new Set()
```

Comment:
We generally use `BufferSet` (ex: https://github.com/iron-fish/ironfish/blob/master/ironfish/src/consensus/verifier.ts#L85) instead of `Set<Buffer>`, but I realize I'm not sure if its necessary or not. I suspect `Set<Buffer>` could suffer from comparison issues that Buffers generally have, though. 

To summarize, we discovered that you can't use basic equality checks with Buffers - they don't always do what one would expect.
Related:
- https://github.com/iron-fish/ironfish/pull/1094
- https://github.com/iron-fish/ironfish/pull/1114
- https://stackoverflow.com/questions/30701220/how-to-compare-buffer-objects-in-nodejs

Reply:
Also right. I actually planned to replace this set with a different thing but then I forgot. Thanks for the pointer to BufferSet--that spared me some time researching alternative. FWIW, the reason why Set is unsuitable is that Buffer is unhashable, so adding Buffers to a Set does not do any "de-duplication" work.